File: app/src/main/java/com/example/lucas/haushaltsmanager/Worker/PeriodicWorker/RecurringBookingWorker.java
package com.example.lucas.haushaltsmanager.Worker.PeriodicWorker;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.WorkerParameters;

import com.example.lucas.haushaltsmanager.Database.AppDatabase;
import com.example.lucas.haushaltsmanager.Database.Repositories.RecurringBookingDAO;
import com.example.lucas.haushaltsmanager.entities.RecurringBooking;
import com.example.lucas.haushaltsmanager.entities.booking.Booking;

import java.util.UUID;

public class RecurringBookingWorker extends AbstractRecurringWorker {
    public static final String RECURRING_BOOKING = "recurringBookingId";
    private static final String TAG = RecurringBookingWorker.class.getSimpleName();

    private final RecurringBookingDAO recurringBookingRepository;

    private final RecurringBooking mRecurringBooking;

    public RecurringBookingWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);

        recurringBookingRepository = AppDatabase.getDatabase(context).recurringBookingDAO();

        UUID recurringBookingId = extractIdFromParameters(workerParams);
        mRecurringBooking = recurringBookingRepository.get(recurringBookingId);
    }

    @NonNull
    @Override
    public Result doWork() {
        if (null == mRecurringBooking) {
            Log.e(TAG, "Terminating RecurringBookingWorker because of invalid RecurringBooking.");
            return Result.failure();
        }

        saveBooking(mRecurringBooking);

        RecurringBooking nextRecurringBooking = RecurringBooking.createNextRecurringBooking(mRecurringBooking);
        if (null != nextRecurringBooking) {
            recurringBookingRepository.update(nextRecurringBooking);

            scheduleNextWorker(nextRecurringBooking);
        } else {
            recurringBookingRepository.delete(mRecurringBooking);
        }

        return Result.success();
    }

    @Override
    String getTag() {
        return "RecurringBookingWorker";
    }

    private void saveBooking(RecurringBooking recurringBooking) {
        AppDatabase.getDatabase(getApplicationContext()).bookingDAO().insert(new Booking(
                UUID.randomUUID(),
                recurringBooking.getTitle(),
                recurringBooking.getPrice(),
                recurringBooking.getDate(),
                recurringBooking.getCategoryId(),
                recurringBooking.getAccountId()
        ));
    }

    private UUID extractIdFromParameters(WorkerParameters workerParams) {
        String rawId = workerParams.getInputData().getString(RECURRING_BOOKING);

        return UUID.fromString(rawId);
    }
}

File: app/src/main/java/com/example/lucas/haushaltsmanager/worker/periodicWorker/RecurringBookingWorker.kt
package com.example.lucas.haushaltsmanager.worker.periodicWorker

import android.content.Context
import android.util.Log
import androidx.work.*
import com.example.lucas.haushaltsmanager.Database.AppDatabase
import com.example.lucas.haushaltsmanager.entities.RecurringBooking
import com.example.lucas.haushaltsmanager.entities.booking.Booking
import java.util.*

class RecurringBookingWorker(context: Context, workerParams: WorkerParameters) : Worker(context, workerParams) {
    companion object {
        private const val WORKER_TAG = "RecurringBookingWorker"

        private const val INPUT_DATA_RECURRING_BOOKING_ID = "recurringBookingId"

        @JvmStatic
        fun createWorkRequest(recurringBooking: RecurringBooking): PeriodicWorkRequest {
            val inputData = workDataOf(
                INPUT_DATA_RECURRING_BOOKING_ID to recurringBooking.id.toString()
            )

            return PeriodicWorkRequestBuilder<RecurringBookingWorker>(recurringBooking.getDelayUntilNextExecution().duration, recurringBooking.getDelayUntilNextExecution().timeUnit)
                .setInputData(inputData)
                .setInitialDelay(recurringBooking.getDelayUntilNextExecution().duration, recurringBooking.getDelayUntilNextExecution().timeUnit)
                .addTag(WORKER_TAG)
                .build()
        }

        fun stopWorker(context: Context) {
            Log.i(WORKER_TAG, "Stopping RecurringBookingWorker")

            WorkManager.getInstance(context).cancelAllWorkByTag(WORKER_TAG)
        }
    }

    override fun doWork(): Result {
        val recurringBookingRepository = AppDatabase.getDatabase(applicationContext).recurringBookingDAO()

        val recurringBookingId = UUID.fromString(inputData.getString(INPUT_DATA_RECURRING_BOOKING_ID))
        val recurringBooking = recurringBookingRepository.get(recurringBookingId)

        saveBooking(recurringBooking)

        val nextRecurringBooking = RecurringBooking.createNextRecurringBooking(recurringBooking)
        if (null == nextRecurringBooking) {
            recurringBookingRepository.delete(recurringBooking)
            stopWorker(applicationContext)
        }

        return Result.success()
    }

    private fun saveBooking(recurringBooking: RecurringBooking) {
        AppDatabase.getDatabase(applicationContext).bookingDAO().insert(
            Booking(
                UUID.randomUUID(),
                recurringBooking.title,
                recurringBooking.price,
                recurringBooking.date,
                recurringBooking.categoryId,
                recurringBooking.accountId,
                null
            )
        )
    }
}